File: frontend/src/api/sync.ts
/**
 * Sync API client.
 */

import apiClient, { API_BASE_URL } from './index';

export type SyncStatus = 'idle' | 'running' | 'completed' | 'failed' | 'cancelled';

export interface SyncStartRequest {
  since_date?: string;
  dry_run?: boolean;
}

export interface SyncStatusResponse {
  status: SyncStatus;
  started_at: string | null;
  progress: number;
  current_account: string | null;
  processed_count: number;
  skipped_count: number;
  error_count: number;
  logs: string[];
}

/**
 * Start a new sync task.
 */
export async function startSync(
  request: SyncStartRequest = {}
): Promise<SyncStatusResponse> {
  const response = await apiClient.post<SyncStatusResponse>('/sync/start', request);
  return response.data;
}

/**
 * Get current sync status.
 */
export async function getSyncStatus(): Promise<SyncStatusResponse> {
  const response = await apiClient.get<SyncStatusResponse>('/sync/status');
  return response.data;
}

/**
 * Cancel the running sync task.
 */
export async function cancelSync(): Promise<{ message: string; cancelled: boolean }> {
  const response = await apiClient.post('/sync/cancel');
  return response.data;
}

/**
 * Create WebSocket connection for real-time sync updates.
 */
export function createSyncWebSocket(
  onMessage: (data: { type: string; message?: string; progress?: number; status?: SyncStatus }) => void,
  onClose?: () => void
): WebSocket {
  const wsUrl = API_BASE_URL.replace(/^http/, 'ws') + '/sync/ws';
  const ws = new WebSocket(wsUrl);

  ws.onmessage = (event) => {
    try {
      const data = JSON.parse(event.data);
      onMessage(data);
    } catch (e) {
      console.error('WebSocket message parse error:', e);
    }
  };

  ws.onclose = () => {
    if (onClose) onClose();
  };

  ws.onerror = (error) => {
    console.error('WebSocket error:', error);
  };

  return ws;
}

File: backend/app/models/sync.py
"""Sync task data models."""

from datetime import datetime
from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, Field


class SyncStatus(str, Enum):
    """Sync task status enumeration."""

    IDLE = "idle"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class SyncStartRequest(BaseModel):
    """Request body for starting a sync task."""

    since_date: Optional[str] = Field(
        None,
        description="Start date for sync, format: '1-JAN-2025'. Defaults to last month.",
        examples=["1-DEC-2024"],
    )
    dry_run: bool = Field(
        False,
        description="If true, only parse without submitting to Firefly III",
    )


class SyncStatusResponse(BaseModel):
    """Response for sync status query."""

    status: SyncStatus = Field(..., description="Current sync status")
    started_at: Optional[datetime] = Field(None, description="Sync start time")
    progress: int = Field(0, ge=0, le=100, description="Progress percentage 0-100")
    current_account: Optional[str] = Field(
        None, description="Currently processing email account"
    )
    processed_count: int = Field(0, description="Number of transactions processed")
    skipped_count: int = Field(0, description="Number of duplicates skipped")
    error_count: int = Field(0, description="Number of errors encountered")
    logs: List[str] = Field(default_factory=list, description="Recent log messages")


class SyncLogMessage(BaseModel):
    """WebSocket message for sync log updates."""

    type: str = Field(..., description="Message type: log/progress/status")
    message: Optional[str] = Field(None, description="Log message content")
    progress: Optional[int] = Field(None, description="Progress percentage")
    status: Optional[SyncStatus] = Field(None, description="Status update")
